devices/blepsynth/blepsynth.cc

```diff
@@ -12,180 +12,293 @@ namespace {
 
 using namespace Ase;
 
-class Envelope
+class ADSR
 {
 public:
-  enum class Shape { EXPONENTIAL, LINEAR };
+  enum class Shape { FLEXIBLE, EXPONENTIAL, LINEAR };
 private:
-  /* values in seconds */
-  float delay_ = 0;
   float attack_ = 0;
-  float hold_ = 0;
+  float attack_slope_ = 0;
   float decay_ = 0;
-  float sustain_ = 0; /* <- percent */
-  float release_ = 0;
-
-  int delay_len_ = 0;
-  int attack_len_ = 0;
-  int hold_len_ = 0;
-  int decay_len_ = 0;
-  int release_len_ = 0;
+  float decay_slope_ = 0;
   float sustain_level_ = 0;
+  float release_ = 0;
+  float release_slope_ = 0;
+  float level_ = 0;
+  float release_start_ = 0;  /* initial level of release stage */
+  int   sustain_steps_ = 0;  /* sustain smoothing */
+  bool  params_changed_ = true;
+  int   rate_ = 48000;
 
-  enum class State { DELAY, ATTACK, HOLD, DECAY, SUSTAIN, RELEASE, DONE };
+  enum class State { ATTACK, DECAY, SUSTAIN, RELEASE, DONE };
 
   State state_ = State::DONE;
-  Shape shape_ = Shape::EXPONENTIAL;
+  Shape shape_ = Shape::LINEAR;
+
+  float a_ = 0;
+  float b_ = 0;
+  float c_ = 0;
 
-  struct SlopeParams {
-    int len;
+  void
+  init_abc (float time_s, float slope)
+  {
+    bool positive = slope > 0;
+    slope = std::abs (slope);
 
-    double factor;
-    double delta;
-    double end;
-  } params_;
+    const float t1y = 0.5f + 0.25f * slope;
 
-  double level_ = 0;
+    a_ = slope * ( 1.0135809670870777f + slope * (-1.2970447050283254f + slope *   7.2390617313972063f));
+    b_ = slope * (-5.8998946320566281f + slope * ( 5.7282487210570903f + slope * -15.525953208626062f));
```

Comment:
I'd be interested to see some performance numbers if it really helps to make the calculations here carried out on floats instead of doubles...

Reply:
I've added a benchmark here:

  https://github.com/tim-janik/dsp-research/blob/trunk/flexadsr/perffloat.cc

Results on my Ryzen 7:

```
ryzen7$ time perffloat float

real    0m0,555s
user    0m0,555s
sys     0m0,000s
ryzen7$ time perffloat double

real    0m0,657s
user    0m0,656s
sys     0m0,001s
```

My experience is that floats are not  faster than doubles, but on modern systems conversion between floats and doubles is. So since we in the inner loop of the ADSR we need floats for a_, b_ and c_, and since if we were modulating these parameters we would likely have a source buffer of floats, it is best to avoid converting to doubles and back for evaluating the polynomials.

Of course it could well be that for modulation we would evaluate the polynomials only every 16 samples (or so) then the conversion cost would not be too bad

Reply:
For reference, on a 11th Gen Intel(R) Core(TM) i7-11800H @ 2.30GHz:
```sh
$ time ./perffloat "float"
real	0m0,364s
user	0m0,364s
$ time ./perffloat "double"         
real	0m0,455s
user	0m0,455s
```